File: 11.container-with-most-water.cpp
/*
 * @lc app=leetcode id=11 lang=cpp
 *
 * [11] Container With Most Water
 */
class Solution
{
public:
	inline int min(int a, int b)
	{
		return a < b ? a : b;
	}
	inline int max(int a, int b)
	{
		return a > b ? a : b;
	}
	int maxArea(vector<int>& height)
	{
		int size = height.size();
		int i = 0, j = size - 1;
		int res = j * min(height[0], height[j]);
		while (i < j)
		{
			height[i] < height[j] ? ++i : --j;
			res = max(res, (j - i) * min(height[i], height[j]));
		}
		return res;
	}
};

File: 11.container-with-most-water.java
/*
 * @lc app=leetcode id=11 lang=java
 *
 * [11] Container With Most Water
 *
 * https://leetcode.com/problems/container-with-most-water/description/
 *
 * algorithms
 * Medium (46.82%)
 * Likes:    5238
 * Dislikes: 542
 * Total Accepted:    586K
 * Total Submissions: 1.2M
 * Testcase Example:  '[1,8,6,2,5,4,8,3,7]'
 *
 * Given n non-negative integers a1, a2, ..., an , where each represents a
 * point at coordinate (i, ai). n vertical lines are drawn such that the two
 * endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together
 * with x-axis forms a container, such that the container contains the most
 * water.
 * 
 * Note: You may not slant the container and n is at least 2.
 * 
 * 
 * 
 * 
 * 
 * The above vertical lines are represented by array [1,8,6,2,5,4,8,3,7]. In
 * this case, the max area of water (blue section) the container can contain is
 * 49. 
 * 
 * 
 * 
 * Example:
 * 
 * 
 * Input: [1,8,6,2,5,4,8,3,7]
 * Output: 49
 */

// @lc code=start
class Solution {
    public int maxArea(int[] height) {
        int i = 0;
        int j = height.length - 1;
        int res = j * Math.min(height[0], height[j]);
        while (i < j) {
            if (height[i] < height[j])
                i++;
            else
                j--;
            int tmp = (j - i) * Math.min(height[i], height[j]);
            if (res < tmp)
                res = tmp;
        }
        return res;
    }
}
// @lc code=end
